### ISSUE TITLE
Make parser more user friendly by accepting bogus content

### ISSUE THREAD

**Issue by JensN4** [NumberFour/n4js#150]
After switching to ES4-like type annotations, the parser produces completely weird results when a Java-like type annotation is used. E.g.

```
export public class A {
    foo(string s, int i) {}
    bar() {}
}
```

produces 7 errors. In particular method bar is not recognized at all. This can also easily happen when forgetting a comma in the parameter list etc.

I suggest to make the parser more robust by introducing "bogus" elements in certain cases. E.g.

```
fragment BindingElementFragment <Yield>*:
    …
    ('=' initializer=AssignmentExpression<In=true, Yield>)?
        bogus=BEFBogus
;

BEFBogus: content+=Identifier*
```

I would assume that this is not possible everywhere, but at least some rules could be made more stable. This would also simplify creating quickfixes, maybe even allowing the conversion of Java-like syntax.
Candidate rules are
- FormalParameter/BindingElementFragment
- MethodDeclaration
- VariableDeclaration
- Class/Interface Declaration

Maybe we could "generalize" the bogus rule, e.g.

```
BEFBogus returns Bogus: BogusFragment<false,false,true,true>;
BogusFragment<comma,braceRight,curlyRight,equals>: (
    ( comma content+=',') 
  | ( braceRight content+=')')
  | ( curlyRight content+='}')
  | (equals content+='=')
  | content+=Identifier
  )*; 

```

**Comment by JensN4** [comment]
Instead of changing the grammar, we should investigate in adding parser customizations to allow for better error recovery and back-tracking. May be postponed until May.

**Comment by meysholdt** [comment]
I've added support for the Java-Style TypeRefs to
- FormalParameter/BindingElementFragment
- CatchVariable
- N4MethodDeclaration
- N4FieldDeclaration
- GetterHeader

But not to VariableDeclaration. See [BogusTypes.n4js.xt](blob/02e89b4b55594ed32c860ae9d2b1b9ddaa98dc82/tests/eu.numberfour.n4js.xpect.tests/model/validation/types/BogusTypes.n4js.xt) for use cases and challenges. 

Regarding the (a) "parser customization" vs. (b) "introduce bogus parser rules" discussion:
(a) can work well adding or removing one or two tokens is sufficient for recovery. The implementation for automatic semicolon insertion is a working example. (b) can handle more complex structures (here: TypeRefs) and also produces an AST subtree for them. That may be helpful to implement quick fixes. 

There was also the idea to add Java-Style TypeRefs to function declarations.

I did not implement recovery for missing commas in parameter lists. That sounds like an opportunity for further parser customization. 

I'll leave this issue open so you guys can decide on potential follow-up issues.
